### PR TITLE
processor/deltatocumulative: telemetry tests

### PR DESCRIPTION
**Pull request by tombrk**
#### Description

Tests internal telemetry (metadata.TelemetryBuilder) is recorded as
expected.

Introduces `internal/testing/sdktest` for this.
Introduces `-- telemetry --` section to testdata.

#### Testing

Existing tests were extended to have a `-- telemetry --` section that specifies expected meter readings in `sdktest.Format`

#### Documentation
not needed

### CODE REVIEW

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (sh0rez) [inline processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: fixes a bug discovered during testing where the callback update had no effect, so the metric was always 0

Comment (RichieSams) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: Are you sure this is correct? `tracked func() int` is a function pointer already. `tracked *func() int` is a pointer to a pointer.

Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: in-fact yes.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e57fcf4154e591d8bcc14f5a00aa796efab9d279/processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go#L20-L26

`trackedCb` captures `m.tracked` by value. if we later re-assign that `func()`, this is not reflected here.
by having it capture a pointer instead, we can update the underlying value and the closure correctly sees the new callback

**Review by ArthurSens — COMMENTED** [review]
I wonder if we couldn't use componentTestTelemetry instead? Testing telemetry is something common to all components in the collector, sounds strange that we need to implement our own comparison logic

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/08a86b491e275092facec75319cb00904e221320/processor/deltatocumulativeprocessor/generated_component_telemetry_test.go#L45-L56

**Comment by ArthurSens** [comment]
```diff
diff --git a/processor/deltatocumulativeprocessor/processor_test.go b/processor/deltatocumulativeprocessor/processor_test.go
index 12d4452e62..cfc28dc617 100644
--- a/processor/deltatocumulativeprocessor/processor_test.go
+++ b/processor/deltatocumulativeprocessor/processor_test.go
@@ -14,11 +14,13 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/config/configtelemetry"
 	"go.opentelemetry.io/collector/confmap/confmaptest"
 	"go.opentelemetry.io/collector/consumer/consumertest"
 	"go.opentelemetry.io/collector/pdata/pmetric"
 	"go.opentelemetry.io/collector/processor"
-	"go.opentelemetry.io/collector/processor/processortest"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/noop"
 	"go.opentelemetry.io/otel/sdk/metric/metricdata"
 	"gopkg.in/yaml.v3"
 
@@ -54,7 +56,7 @@ func TestProcessor(t *testing.T) {
 
 			ctx := context.Background()
 			cfg := config(t, file("config.yaml"))
-			proc, sink := setup(t, cfg)
+			proc, sink, tel := setup(t, cfg)
 
 			stages, _ := filepath.Glob(file("*.test"))
 			for _, file := range stages {
@@ -71,6 +73,8 @@ func TestProcessor(t *testing.T) {
 					t.Fatal(diff)
 				}
 			}
+
+			tel.assertMetrics(t, metrics)
 		})
 
 	}
@@ -89,7 +93,7 @@ func config(t *testing.T, file string) *Config {
 	return cfg
 }
 
-func setup(t *testing.T, cfg *Config) (processor.Metrics, *consumertest.MetricsSink) {
+func setup(t *testing.T, cfg *Config) (processor.Metrics, *consumertest.MetricsSink, componentTestTelemetry) {
 	t.Helper()
 
 	next := &consumertest.MetricsSink{}
@@ -97,15 +101,27 @@ func setup(t *testing.T, cfg *Config) (processor.Metrics, *consumertest.MetricsS
 		cfg = &Config{MaxStale: 0, MaxStreams: math.MaxInt}
 	}
 
+	tel := setupTestTelemetry()
+	set := tel.NewSettings()
+	mp := set.LeveledMeterProvider(configtelemetry.LevelBasic)
+	set.LeveledMeterProvider = func(level configtelemetry.Level) metric.MeterProvider {
+		// detailed level enables otelhttp client instrumentation which we
+		// dont want to test here
+		if level == configtelemetry.LevelDetailed {
+			return noop.MeterProvider{}
+		}
+		return mp
+	}
+
 	proc, err := NewFactory().CreateMetrics(
 		context.Background(),
-		processortest.NewNopSettings(),
+		set,
 		cfg,
 		next,
 	)
 	require.NoError(t, err)
 
-	return proc, next
+	return proc, next, tel
 }
```

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (ArthurSens) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: Would it make sense to use separate commits here? One that adds the test, even if it fails, another that fixes a bug.

If the there is a test failing before the fix it would be awesome since it show that it is indeed a bug :)

File: processor/deltatocumulativeprocessor/internal/testing/sdktest/example_test.go
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/testing/sdktest/example_test.go:79]: nope.

`go-cmp` has unstable output (by design, they alter the type of whitespace they use on each invocation so nobody relies on stable output)

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: sure

**Comment by sh0rez** [comment]
@ArthurSens I considered `assertMetrics` and felt it ultimately does not do what we need.

First of all, it requires you to specify _all_ metrics emitted by the reader, which we rarely want. We only want to test the metrics relevant to the specific test case.

Second, you need _some_ way to get that `expected` slice, and the most of `sdktest` implements exactly that: a format to specify the expected slice.

For actual comparison, it just calls out to `go-cmp` which imo yields way more usable results that `assertMetrics` does:

- **assertMetrics**:
	```
    processor_test.go:80: [Metrics Data not equal: Sum[int64] not equal: Sum DataPoints not equal:
        missing expected values:
        metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[1]attribute.KeyValue{attribute.KeyValue{Key:"error", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"delta.ErrOlderStart", slice:interface {}(nil)}}}}}, StartTime:time.Date(2024, time.November, 1, 18, 54, 31, 646527291, time.Local), Time:time.Date(2024, time.November, 1, 18, 54, 31, 647399262, time.Local), Value:2, Exemplars:[]metricdata.Exemplar[int64]{}}
        unexpected additional values:
        metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[1]attribute.KeyValue{attribute.KeyValue{Key:"error", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"delta.ErrOlderStart", slice:interface {}(nil)}}}}}, StartTime:time.Date(2024, time.November, 1, 18, 54, 31, 646527291, time.Local), Time:time.Date(2024, time.November, 1, 18, 54, 31, 647421097, time.Local), Value:1, Exemplars:[]metricdata.Exemplar[int64]{}}
        ]
	```
- **sdktest**:
	```
    processor_test.go:83:
          []metricdata.Metrics{
                {
                        Name:        "otelcol_deltatocumulative.datapoints.linear",
                        Description: "total number of datapoints processed. may have 'error' attribute"...,
                        Unit:        "{datapoint}",
                        Data: metricdata.Sum[int64]{
                                DataPoints: []metricdata.DataPoint[int64](Inverse(sdktest.Transform.int64, []sdktest.DataPoint[int64]{
                                        {DataPoint: {StartTime: s"2024-11-01 18:54:31.646527291 +0100 CET m=+0.020564831", Time: s"2024-11-01 18:54:31.647621443 +0100 CET m=+0.021658982", Value: 3, Exemplars: {}, ...}, Attributes: {}},
                                        {
                                                DataPoint: metricdata.DataPoint[int64]{
                                                        ... // 1 ignored field
                                                        StartTime: s"2024-11-01 18:54:31.646527291 +0100 CET m=+0.020564831",
                                                        Time:      s"2024-11-01 18:54:31.647621443 +0100 CET m=+0.021658982",
        -                                               Value:     2,
        +                                               Value:     1,
                                                        Exemplars: {},
                                                },
                                                Attributes: {"error": string("delta.ErrOlderStart")},
                                        },
                                        {DataPoint: {StartTime: s"2024-11-01 18:54:31.646527291 +0100 CET m=+0.020564831", Time: s"2024-11-01 18:54:31.647621443 +0100 CET m=+0.021658982", Value: 1, Exemplars: {}, ...}, Attributes: {"error": string("delta.ErrOutOfOrder")}},
                                })),
                                Temporality: s"CumulativeTemporality",
                                IsMonotonic: true,
                        },
                },
          }
```

I can honestly not even spot the difference in the assertMetrics output


Taking a closer look, I think I can cut down the amount of work sdktest does by relying more on go-cmp ignorers ... similar to how `assertMetrics` is implemented with those `Ignore*` options

**Comment by ArthurSens** [comment]
> First of all, it requires you to specify _all_ metrics emitted by the reader, which we rarely want. We only want to test the metrics relevant to the specific test case.

Could we contribute this?

I'd rather stick to what everybody is using, and improve things for everybody if we see limitations

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (RichieSams) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:40]: Interesting.... I would have thought that go would have captured `m` and called its method. But it's capturing `m.tracked` directly?

**Review by ArthurSens — APPROVED** [review]
Hey @RichieSams, I am just summarizing here what @sh0rez and I discussed in Slack:

Although we're writing a lot of code here that concerns our test framework, the PR does test that the telemetry behaves as expected, so let's merge this to unblock other work.

We did agree that we don't appreciate maintaining a full-blown test framework just for us. It's extra work for the maintainers of this component and we don't help the whole collector community evolve what they have.

Task for the future:
* Propose better test output for collector-core's testing framework.

File: processor/deltatocumulativeprocessor/testdata/exponential/1.test
Comment (MovieStoreGuy) [inline processor/deltatocumulativeprocessor/testdata/exponential/1.test:1]: Any reason why this isn't a `.yml` or `.yaml` file?

Comment (sh0rez) [reply processor/deltatocumulativeprocessor/testdata/exponential/1.test:1]: it's technically not a yaml file.

this uses [txtar](https://pkg.go.dev/golang.org/x/tools/txtar) to bundle several yaml files (in, out, telemetry) into one physical file. This is nice because:
- input and output in the same file
- multiple write-requests per testcase as each request is a `.test` file. This allows to test processor state across write requests

the `.test` ending is arbitrary but I thought it serves the purpose here

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (jpkrohling) [inline processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:19]: Am I the only one who thinks "zero" is a bad name for something that returns `-1`? Also because the previous iteration of the code returned zero...

**Review by jpkrohling — APPROVED** [review]

File: processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go
Comment (sh0rez) [reply processor/deltatocumulativeprocessor/internal/lineartelemetry/metrics.go:19]: ugh this is an oversight.
I will address this in a later PR

**Pull request merged**
